### DISCUSSION TITLE
Automatic Metadata download for data readers

### DISCUSSION THREAD

**Discussion by javihern98** [bis-med-it/pysdmx#151] [Ideas]
Implement the automatic download of metadata (mainly Schema) whenever we read a Dataset. If we find a Schema to download from the APIs, we will set Dataset.structure to it, if not we leave the structure reference (as string)

@sosna @stratosn  I need some help to specify this:

- Which methods shall we use? (We needed the GDS API for this in production, maybe we can find an alternative)
  - Shall we use the Query builder to get the APIs from the different agencies?
  - Read some environment variable for the FMR instance from which we can download the data? (using [SDMX Global Registry](https://registry.sdmx.org/) as default?) and use the Fusion API
  - Any other alternative using a standard SDMX REST API?

**Comment by sosna** [comment]
Just to be sure I understood correctly: A dataset message contains a reference (for example, as a URN) to one or more structures (data structure or dataflow or provision agreement) to which the data in the message relate. Is your question about how we can translate this URL into a REST request to fetch the structures?

If so, I would say that the most elegant way would be to use the SGDS URN resolution mechanism to know the authoritative service from which the structure can be fetched.

If the SGDS URN resolution mechanism returns the entry point of the service (e.g. https://registry.sdmx.org/sdmx/v2/), then I would build the query using the query builder indeed.

If the SGDS URN resolution mechanism returns the full URL (e.g.: https://registry.sdmx.org/sdmx/v2/structure/dataflow/ESTAT/BPM6_BOP_M/2.3), then I think we should parse it to build the appropriate schema query, again using the query builder, as this is probably the best artefact for validation purposes.

What do you think?
